src/common.ts

```diff
@@ -1,84 +1,74 @@
+import { atom } from 'jotai/vanilla'
 import type { Source } from 'wonka'
-
-export const createAtoms = <
-  Args,
-  Result extends OperationResult,
-  Action,
-  ActionResult
->(
+import { pipe, toObservable } from 'wonka'
+import { suspenseAtom } from './suspenseAtom'
+
+export type InitialOperationResult<Data, Variables extends AnyVariables> = Omit<
+  OperationResult<Data, Variables>,
+  'operation'
+> & {
+  operation: Operation<Data, Variables> | undefined
+}
+// This is the same (aside from missing fetching and having hasNext) object shape as urql-react has by default while operation is yet to be triggered/yet to be fetched
+export const urqlReactCompatibleInitialState = {
+  stale: false,
+  // Casting is needed to make typescript chill here as it tries here to be too smart
+  error: undefined as any,
+  data: undefined as any,
+  extensions: undefined as any,
+  hasNext: false,
+  operation: undefined,
+} as InitialOperationResult<any, any>
+
+export const createAtoms = <Args, Result extends OperationResult, ActionResult>(
   getArgs: (get: Getter) => Args,
   getClient: (get: Getter) => Client,
   execute: (client: Client, args: Args) => Source<Result>,
-  handleAction: (
-    action: Action,
-    client: Client,
-    refresh: () => void
-  ) => ActionResult
+  reexecute: (context: Partial<OperationContext>, get: Getter) => ActionResult,
+  getPause: (get: Getter) => boolean
 ) => {
-  const refreshAtom = atom(0)
-
-  if (process.env.NODE_ENV !== 'production') {
-    refreshAtom.debugPrivate = true
-  }
-
   const sourceAtom = atom((get) => {
-    get(refreshAtom)
     const args = getArgs(get)
     const client = getClient(get)
-    const source = execute(client, args)
+    const source = getPause(get) ? null : execute(client, args)
     return source
   })
 
   if (process.env.NODE_ENV !== 'production') {
     sourceAtom.debugPrivate = true
   }
 
-  const baseStatusAtom = atom((get) => {
-    const source = get(sourceAtom) as Source<Result | undefined>
-    const observable = pipe(source, toObservable)
-    const resultAtom = atomWithObservable(() => observable, {
-      initialValue: undefined,
-    })
-
-    if (process.env.NODE_ENV !== 'production') {
-      resultAtom.debugPrivate = true
-    }
+  const initialLoadAtom = atom<Result>(
+    urqlReactCompatibleInitialState as Result
+  )
 
-    return resultAtom
+  const cacheResultAtom = atom<{ cache: Map<number, Result> }>({
+    cache: new Map(),
   })
-
-  if (process.env.NODE_ENV !== 'production') {
-    baseStatusAtom.debugPrivate = true
-  }
-
-  const statusAtom = atom(
-    (get) => {
-      const resultAtom = get(baseStatusAtom)
-      return get(resultAtom)
-    },
-    (get, set, action: Action) => {
-      const client = getClient(get)
-      const refresh = () => {
-        set(refreshAtom, (c) => c + 1)
-      }
-      return handleAction(action, client, refresh)
+  cacheResultAtom.onMount = (setAtom) => {
+    return () => {
+      setAtom({ cache: new Map() })
```

Comment:
So, you should create a new cache on mount. Not on unmount.

Reply:
Here I'm resetting it. It's already a `new Map()` on creation. Here I just cleared away the cache when a person will redirect out of the page that uses the query. It's for `pause` functionality only.